mlir/lib/Dialect/Quant/Transforms/StripFuncQuantTypes.cpp

```diff
@@ -0,0 +1,114 @@
+//===- StripFuncQuantTypes.cpp - Strip quantized types --------------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// Strips quantized types from function headers.
+//
+//===----------------------------------------------------------------------===//
+
+#include "mlir/Dialect/Arith/IR/Arith.h"
+#include "mlir/Dialect/Func/IR/FuncOps.h"
+#include "mlir/Dialect/Func/Transforms/FuncConversions.h"
+#include "mlir/Dialect/Linalg/IR/Linalg.h"
+#include "mlir/Dialect/Quant/IR/Quant.h"
+#include "mlir/Dialect/Quant/IR/QuantTypes.h"
+#include "mlir/Dialect/Quant/Transforms/Passes.h"
+#include "mlir/Dialect/Shape/IR/Shape.h"
+#include "mlir/Dialect/Tensor/IR/Tensor.h"
+#include "mlir/IR/Matchers.h"
+#include "mlir/IR/PatternMatch.h"
+#include "mlir/Transforms/DialectConversion.h"
+
+namespace mlir {
+namespace quant {
+
+#define GEN_PASS_DEF_STRIPFUNCQUANTTYPES
+#include "mlir/Dialect/Quant/Transforms/Passes.h.inc"
+
+namespace {
+
+class QuantizedTypeConverter : public TypeConverter {
+
+  static Type convertQuantizedType(QuantizedType quantizedType) {
+    return quantizedType.getStorageType();
+  }
+  
+  static Type convertTensorType(TensorType tensorType) {
```

Comment:
The original TableGen spec of the ops supported both tensor and vector types. I decided to avoid the vector type altogether in an effort to simplify this initial implementation. Adding support for vectors or memrefs could be a follow-up task, which would involve carefully revisiting all contexts in which quantized tensors can occur, as these containers may need special treatments in some cases. For example, does it make sense to consider per-channel quantization in a vector? For now, and as you suggest, I don't see a specific use case, assuming it is reasonable for quantization lowering to occur before bufferization and vectorization.

Reply:
Are you referring to per-channel quantization in vectors specifically, Jacques? I regard a vector as a low-level data bundle where arithmetic can occur homogeneously across its elements. If a given vector dimension contains channels with different quantization information, we'd need to first normalize it before operating on it. Not impossible, but I can't see why we would need that, considering that quantized types are likely lowered at the tensor level, before bufferization and vectorization.